[driver] driver for ST vl53l5, vl53l7, vl53l8

hshose opened the pull request:
## STs Ultra Light Driver for VL53L5 (I2C), VL53L7 (I2C), and VL53L8 (I2C, SPI) 8x8 Pixel Time-of-Flight Sensors

As the interaction with the sensor is pretty complicated (many undocumented registers and huge binary blobs), I use the ultra light drivers (ULD) for the sensors as provided by ST.
If you think this is a good way of integrating support for these sensors, I suggest to add these ULDs as a submodule under ext (see temp commit here for proposed structure). Does this work for y'all?
The VL53L5 driver exists undre STs namespace on github, the VL53LMZ driver for VL53L7 and VL53L8 only on STs website, I thus decided on STs website as datasorce for all drivers; I hope this is fine?
The `update.py` script in the future submodule automatically extracts the files, fixes some typos and bugs in their C-API, prefixes global functions with "VL53_", and unifies the use of the `VL53_Platform` struct among both drivers and their examples.

## modm transport
I added the modm transport SPI and I2C classes in the `ext/vl53` folder, is this the right place or should they go in the `src/modm/driver/position` folder?

## Examples
The VL53L8 sensor supports SPI and I2C, the VL53L7 sensor (same driver as the VL53L8 but wider field of view) supports only I2C, the VL53L5 supports only I2C.
Examples should cover these cases. 

## Todos:
- [x] Test VL53L8 SPI in hardware
- [x]  Test VL53L5 I2C in hardware
- [x] Test VL53L7/VL53L8 I2C in hardware

## Future plans:

Add support for the upcoming [VL53L9](https://www.st.com/en/imaging-and-photonics-solutions/vl53l9cx.html) when its released

chris-durand commented on ext/vl53/vl53_transport.hpp:
```diff
@@ -0,0 +1,140 @@
+	virtual ~Vl53TransportBase() = default;
+
+	virtual uint8_t
+	writeMulti(uint16_t reg, uint8_t *values, uint32_t size) = 0;
```
```suggestion
	writeMulti(uint16_t reg, const uint8_t *values, uint32_t size) = 0;
```

It doesn't hurt to make it `const`-correct even if the calling code doesn't ask for it.

hshose replied:
fixed

chris-durand commented on ext/vl53/vl53_transport.hpp:
```diff
@@ -0,0 +1,140 @@
+
+	virtual uint8_t
+	writeMulti(uint16_t reg, uint8_t *values, uint32_t size) = 0;
+	virtual uint8_t
```
I presume the return values should correspond to the C API? Your I2C implementation will return 1 on success, but the SPI transport always returns 0. That doesn't match ;)

hshose replied:
whuuuuupsiii :-D You're absolutely right!

chris-durand commented on ext/vl53/vl53_transport.hpp:
```diff
@@ -0,0 +1,140 @@
+	uint8_t
+	readMulti(uint16_t register_address, uint8_t *p_values, uint32_t size) override;
+
+	inline uint8_t
```
```suggestion
	uint8_t
```
No explicit `inline` required here. It's implicitly `inline` because it's a definition inside the class declaration body.

hshose replied:
fixed

chris-durand commented on ext/vl53/vl53_transport_impl.hpp:
```diff
@@ -0,0 +1,191 @@
+Vl53SpiTransport<SpiMaster, Cs, Lpn, VL53LMZ_COMMS_CHUNK_SIZE>::resetSensor()
+{
+	Lpn::reset();
+	modm::delay_ms(100);
```
This should use some fiber wait mechanism instead of a blocking delay.

hshose replied:
fixed.

chris-durand commented on ext/vl53/vl53_transport_impl.hpp:
```diff
@@ -0,0 +1,191 @@
+	uint8_t status = 0;
+	uint32_t position = 0;
+	uint32_t data_size = 0;
+	uint16_t temp;
```
I'm guessing this was adapted from some C89 code? `position` can go into to the `for` statement and `temp` can be defined in the body for better readability.

hshose replied:
Yes. I fixed it as you suggested in both SPI functions.

chris-durand reviewed (commented):
Very nice! I don't have the time for a full review because I'll be on a dive boat in the Andaman sea in 2 hours ;)

Putting ST's code into ext is the right way to go. Getting it from the website is also fine because it's BSD licensed and you are providing the automated update script. @salkinium or @rleh can probably help with making a repo in modm-ext.

chris-durand commented on ext/vl53/vl53_transport_impl.hpp:
```diff
@@ -0,0 +1,191 @@
+		std::memcpy(p_values + position, &rx_buffer[2], data_size);
+	}
+
+	return status;
```
`status` is always 0

hshose replied:
Not sure if SPI has a status that is returned. I now just always return 0 directly and avoid the unused status variable.

hshose commented:
All examples work in HW now. If you set up the ext repo, I can add the submodule and move the ST files there...

rleh commented:
I created https://github.com/modm-ext/vl53-partial and granted you access.

hshose commented:
Adding submodule done.

salkinium commented on .gitmodules:
```diff
@@ -52,3 +52,6 @@
 	url = git@github.com:modm-ext/segger-rtt.git
+[submodule "ext/vl53/vl53"]
+	path = ext/vl53/vl53
+	url = git@github.com:modm-ext/vl53-partial.git
```
Could you change this URL to the HTTPS method? Otherwise it will fail for people without a GitHub SSH key setup. See https://github.com/modm-io/modm/pull/1297.

hshose replied:
sure, fixed this.

salkinium commented on examples/nucleo_h723zg/vl53/vl53l5cx_i2c/project.xml:
```diff
@@ -0,0 +1,15 @@
+  <options>
+    <option name="modm:build:build.path">../../../../build/nucleo_h723zg/vl53l5cx_i2c</option>
+    <option name="modm:vl53:vl53l5cx:examples">yes</option>
+    <option name="modm:platform:cortex-m:main_stack_size">64Ki</option>
```
WAT? is this a left over of the development, or do the examples really use 64kB of stack?

hshose replied:
The I2C version needs the 64kB (or at least it stops working if I set 32kB stack, not really sure how to debug stack usage in modm.
The SPI version needs only 16kB (depending on the size of the buffer for the SPI read/write).

The firmware needs a bunch of large buffers for the transfers and also has pretty large firmware blobs (`cost uint8_t[LARGE]`), not sure where this breaks with smaller stack.

salkinium replied:
It's because everything in the example functions is allocated on the stack.

- sizeof(VL53L5CX_ResultsData) = 1360
- sizeof(VL53L5CX_Configuration) = 2744

But that's just 4KB together, where does the other 60kB come from?
I'm worried there's some kind of recursion inside the drivers that the port is causing.
Did you test the driver outside of the examples yet?

salkinium replied:
Ah, found it. You allocate the I2C write buffer on the stack here:
```cpp
template<typename I2cMaster, typename Lpn>
uint8_t
Vl53I2cTransport<I2cMaster, Lpn>::writeMulti(uint16_t RegisterAddress, const uint8_t *p_values,
											 uint32_t size)
{
	uint8_t data_write[size + 2];
	data_write[0] = (RegisterAddress >> 8) & 0xFF;
	data_write[1] = RegisterAddress & 0xFF;
```
and this gets called in the `vl53l5cx_init` function with size = 0x8000, which is 32kB.
```cpp
	status |= VL53_WrMulti(&(p_dev->platform),0,
		(uint8_t*)&VL53L5CX_FIRMWARE[0x8000],0x8000);
```

You should instead inherit privately from [the `modm::I2cEeprom` class](https://github.com/modm-io/modm/blob/develop/src/modm/driver/storage/i2c_eeprom.hpp#L70), which allows for copy-less prefixing of a 2-byte address.

hshose replied:
This is nice! Didn't know to look in the storage drivers for this. Gets the example Stack down to like 12kB.

salkinium commented on ext/vl53/vl53.hpp:
```diff
@@ -0,0 +1,73 @@
+uint8_t
+VL53_WaitMs(VL53_Platform *p_platform, uint32_t TimeMs)
+{
+	modm::delay_ms(TimeMs);
```
```suggestion
	modm::this_fiber::sleep_for(std::chrono::milliseconds(TimeMs));
```

hshose replied:
fixed

hshose replied:
fixed

salkinium commented on ext/vl53/vl53.lb:
```diff
@@ -0,0 +1,86 @@
+        env.copy("vl53/vl53l5cx/inc")
+        env.copy("vl53/vl53l5cx/LICENSE.txt")
+        if env[":vl53:vl53l5cx:examples"]:
+            env.copy("vl53/vl53l5cx/examples")
```
I weakly feel like this shouldn't be an lbuild option.
At least it's not consistent the the CMSIS-DSP examples, which are [instead included from the source path](https://github.com/modm-io/modm/blob/develop/examples/nucleo_f429zi/cmsis_dsp/signal_converge/main.cpp).

hshose replied:
Really don't have any strong opinion on this.
I thought the point of lbuild was to copy the files you need from the top-level modm directory to your local project directory and then using macros in the examples to access ext files seems to contradict that paradigm.

Anyways, I changed it to be consistent with CMSIS DSP.

salkinium commented on ext/vl53/vl53.lb:
```diff
@@ -0,0 +1,86 @@
+    return True
+
+def build(env):
+    env.outbasepath = "modm/ext/vl53"
```
```suggestion
    env.outbasepath = "modm/src/modm/driver/vl53"
```
To separate the modm part from the ext part.

hshose replied:
done

salkinium commented on ext/vl53/vl53_platform.h:
```diff
@@ -0,0 +1,219 @@
+#ifndef VL53_PLATFORM_H_
+#define VL53_PLATFORM_H_
+#pragma once
+
```
```suggestion
#if __has_include("modm_config_vl53.h")
#include "modm_config_vl53.h"
#endif
```
To allow user configuration without modifying this library file.

hshose replied:
done

salkinium reviewed (changes requested):
Thanks, just some minor improvements needed.

hshose commented:
Thank you @salkinium for the detailed review, I tried to address most of the issues, only the stack and example include questions remain. I left the stack as is because the I2C examples seem to need this, I changed the SPI examples stack to 16kB.
Changed the examples as separate commit in case you don't want to have this.

salkinium commented on ext/vl53/vl53_transport.hpp:
```diff
@@ -0,0 +1,140 @@
+	inline void
+	SetAddress(const uint8_t address)
+	{
+		this->address_ = address;
```
Isn't this functionality already [part of `modm::I2cDevice`](https://github.com/modm-io/modm/blob/develop/src/modm/architecture/interface/i2c_device.hpp#L58)?

hshose replied:
whups, okay yes!

salkinium commented on ext/vl53/vl53_transport_impl.hpp:
```diff
@@ -0,0 +1,182 @@
+
+		modm::this_fiber::poll([&] { return this->acquireMaster(); });
+		Cs::reset();
+		SpiMaster::transfer(&tx_buffer[0], nullptr, data_size);
```
Why not do it in two transfers?
```cpp
uint8_t addr_buffer[2];
addr_buffer[0] = SPI_WRITE_MASK(temp) >> 8;
addr_buffer[1] = SPI_WRITE_MASK(temp) & 0xFF;
SpiMaster::transfer(addr_buffer, nullptr, sizeof(addr_buffer));
SpiMaster::transfer(p_values, nullptr, size);
```
that saves the memcpy and the scratch buffer

hshose replied:
I used to do this, and it works with the `SpiMaster2`.
However, with the `SpiMaster2_Dma` it doesn't work and I don't really know why. Any ideas?

When I have the separate TX Buffer and memcopy the write content into that Buffer, it also works with the `SpiMaster2_Dma`.

salkinium replied:
Not sure, but the DMA buffers need to be 4B aligned. If you make the `addr_buffer[0]` a member it should be 4B aligned.

hshose replied:
Making addr_buffer a member doesn't help. How would i know that `p_values` is aligned properly?

chris-durand replied:
> Not sure, but the DMA buffers need to be 4B aligned. If you make the `addr_buffer[0]` a member it should be 4B aligned.

Is that really required? The SPI DMA only performs byte accesses and it should even support unaligned access on normal memory. The only references on DMA alignment requirements I could find were on device and strongly-ordered memory which need to be properly aligned.

However, when you use cacheable memory DMA buffers need to be aligned to the cache line size for cache maintenance operations to sensibly work. modm doesn't support that anyway. @hshose Just to make sure, you didn't manually enable the data cache, right?

hshose replied:
I wouldn't even know how to turn this on :smile: Also making the `addr_buffer` a class member doesn't help.

salkinium replied:
I pushed a workaround to a common cache bug on Cortex-M7. It probably won't fix the DMA issue, but it's something we need to fix in modm anyways.

chris-durand replied:
Erratum 1259864 only applies if a memory region is configured in write-through mode. In the default memory map all internal cacheable RAM and almost all external RAM is write-back write-allocate. At least in the 3 H7 manuals I had on my laptop. Only external RAM connected via QUAD/OCTOSPI would be write-through by default. We don't support that, so we should be fine.

<img width="786" height="621" alt="Screenshot_20251121_173045" src="https://github.com/user-attachments/assets/65f883b7-2a1a-490e-96ac-5c034c9d4fbf" />

salkinium replied:
Yup I know, but somehow this fixed several issues I was seeing on a H743/r1p1. Maybe this helps here. Otherwise it's a bug in the SPI DMA implementation.

chris-durand replied:
Would be really interesting if it made any difference here because the data cache should be off and the examples are on an H723/r1p2 which shouldn't have this issue.

chris-durand replied:
@hshose Could you check with a scope or logic analyzer what goes wrong if you split the transfers with DMA?

salkinium replied:
> are on an H723/r1p2

Oh, then my "fix" is indeed useless. 😞

hshose replied:
I am not entirely sure what if or what changed but now that I'm looking with a logic analyzer it also works with the DMA :facepalm: I should go home and do something else with my life :crying_cat_face:

hshose replied:
Maybe it was the alignement issue and I forgot to invoke lbuild on when I tested :see_no_evil: 
Oh well.

hshose replied:
![image](https://github.com/user-attachments/assets/ada1f8ca-13b6-4c4b-9daa-e16304c80912)

Also the testing setup is a bit of a mess.

salkinium commented on ext/vl53/vl53_transport.hpp:
```diff
@@ -0,0 +1,138 @@
+};
+
+template<typename I2cMaster, typename Lpn>
+class Vl53I2cTransport : public modm::I2cDevice<I2cMaster>, public Vl53TransportBase, private I2cEeprom<I2cMaster, 2>
```
```suggestion
class Vl53I2cTransport : public modm::I2cEeprom<I2cMaster, 2>, public Vl53TransportBase
```

salkinium commented on ext/vl53/vl53_transport.hpp:
```diff
@@ -0,0 +1,138 @@
+class Vl53I2cTransport : public modm::I2cDevice<I2cMaster>, public Vl53TransportBase, private I2cEeprom<I2cMaster, 2>
+{
+public:
+	Vl53I2cTransport(const uint8_t address) : I2cDevice<I2cMaster>{address}, I2cEeprom<I2cMaster,2>{address} {};
```
```suggestion
	Vl53I2cTransport(const uint8_t address) : I2cEeprom<I2cMaster,2>{address} {};
```

salkinium commented on ext/vl53/vl53_transport.hpp:
```diff
@@ -0,0 +1,138 @@
+	{
+		I2cEeprom<I2cMaster,2>::transaction.setAddress(address);
+		I2cDevice<I2cMaster>::transaction.setAddress(address);
+	};
```
this function is already implemented by I2cDevice

hshose replied:
wasn't accessible before with the private I2CEEprom... but is now with the other changes.

salkinium commented on ext/vl53/vl53_transport_impl.hpp:
```diff
@@ -0,0 +1,174 @@
+	data_write[0] = (RegisterAddress >> 8) & 0xFF;
+	data_write[1] = RegisterAddress & 0xFF;
+
+	const auto status = I2cDevice<I2cMaster>::writeRead(&data_write[0], 2, &data_read[0], 1);
```
```suggestion
	const auto status = I2cDevice<I2cMaster>::writeRead(&data_write[0], 2, p_value, 1);
```

salkinium commented:
Yes, the macOS CI is super slow (~2h), I'm trying to fix it over at #1301.

salkinium commented on ext/vl53/vl53_transport_impl.hpp:
```diff
@@ -0,0 +1,164 @@
+		Cs::reset();
+		// SpiMaster::transfer(&tx_buffer[0], nullptr, data_size);
+		SpiMaster::transfer(addr_buffer, nullptr, sizeof(addr_buffer));
+		SpiMaster::transfer(&p_values[position], nullptr, data_size);
```
Dumb question: Why are we chunking this here? Is there a limit on the SPI transfer length? Cos the I2C just dumps 32kB of data into the sensor in one transfer, if I understand this correctly.

hshose replied:
The ST ULD implementation does this and due to the lack of documentation for the sensors besides the provided drivers, I just did the same.
I can test tomorrow if we can just write the entire firmware blob at once.

salkinium replied:
Ping? I didn't find a good way to download the libraries quickly. But I can fix it later, it shouldn't block this PR.

hshose replied:
```
64 bytes from hshose (1234:5678:7654:321::2003): icmp_seq=1 ttl=117 time=54345123.19 ms
```
Sorry, my dog ate my homework :poodle: 

It also works in hardware without the chunking (in both read and write).

salkinium reviewed (commented):
I like this, thank you! I will have a look at the submodule until Monday and check if I can get it to update automatically by going past the login wall like we do for the CubeMX database.

hshose committed:
[driver] Add vl53l5, vl53l7, vl53l8 drivers

hshose committed:
[examples] Add VL53x examples for I2C and SPI

salkinium commented:
I squashed your commits into two and then added two fixup commits with my changes. Mostly just moving files into the right place and reducing stack usage to 8kB. Could you test that again please and then squash the fixups if it works?

hshose commented:
8k stack is too little for the VL53L7 or L8. it works with 12k.

For the VL53L5 8K is enough.

I have updated the examples accordingly
